posts/how-to-get-complete-leverage-from-javascript-reduce/2019-06-28-how-to-get-complete-leverage-from-js-reduce.md

```diff
@@ -0,0 +1,394 @@
+
+```javascript
+const taskWithCompleteFlag = boolenFlag => task => task.completed === boolenFlag
+const taskCreatedWithinDays = numberOfDays => task =>
+  moment(task.createdAt).diff(moment().subtract(numberOfDays, "days"), "days") >
+  0
+const taskWithHigherPriority = priorityLevel => task =>
+  task.priorityLevel < priorityLevel
+const taskWithAnyTags = tags => task =>
+  tags.reduce((boolAcc, tag) => boolAcc || task.tags.includes(tag), false)
+
+const capitalizeWord = word => word.charAt(0).toUpperCase() + word.slice(1)
+const capitalizeEachWord = template =>
+  template
+    .split(" ")
+    .map(capitalizeWord)
+    .join(" ")
+
+const capitalizeTaskDescription = task => ({
+  ...task,
+  desc: capitalizeEachWord(task.desc),
+})
+
+const tagsGrouper = (acc, task) => {
+  task.tags.forEach(tag => {
+    acc[tag] = task
+  })
+
+  return acc
+}
+
+const filteredTasks = tasks
+  .filter(taskCreatedWithinDays(7))
+  .filter(taskWithCompleteFlag(true))
+  .filter(taskWithHigherPriority(3))
+  .filter(taskWithAnyTags(["Work", "Personal"]))
+  .map(capitalizeTaskDescription)
+  .reduce(tagsGrouper, {})
+
+console.log(filteredTasks)
+```
+
+If the report needs to include tasks created within the last two weeks, filter call will be taskCreatedWithinDays(14) instead oftaskCreatedWithinDays(7).
+
+By moving out the filter attributes from the function definition to the function arguments, filter logic is made completely configurable.
+
+Assume that the report logic remains the same except that the tasks created within the last 31 days should be considered instead of 7. A solution can be designed similar to below snippet.
+
+```javascript
+const getCommonReportTasks = tasks =>
+  tasks
+    .filter(taskWithCompleteFlag(true))
+    .filter(taskWithHigherPriority(3))
+    .filter(taskWithAnyTags(["Work", "Personal"]))
+
+const monthlyReportTasks = getCommonReportTasks(tasks)
+  .filter(taskCreatedWithinDays(31))
+  .map(capitalizeTaskDescription)
+  .reduce(tagsGrouper, {})
+
+console.log(monthlyReportTasks)
+```
+
+What about the problem statement alongside the following changes?
+
+- Tasks which are incomplete instead of completed
+- Tasks created within the last 2 months instead of last one week/month
+- Priority Levels can be ignored
+
+Solving these problems using this approach will result in a lot of redundant functions with slight changes. _Even though this approach is completely configurable, it’s not composable (combining functions to create a new function)._
+
+## Function Pipelining to the rescue
+
+Can you infer any traces of reduce()functionality in the last approach?
+
+The tasks is an array which is processed by a series of functions. Each function (accumulator callback) in the series is filtering the array and finally, the array is transformed into an object (tags-grouped tasks).
+
+### pipe()
+
+```javascript
+const pipe = (...functions) => (value) => (
+  **functions.reduce**((currValue, currFunc) => (
```

Comment:
Is these ** as really intended? We dont need to write any custom styling inside code block as code block has its own syntactical styles applied.

Reply:
Removed emphasizing